examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,605 @@
+    # Create the dual encoder model.
+    dual_encoder = keras.Model(
+        inputs=[captions, images], outputs=outputs, name="dual_encoder"
+    )
+    return dual_encoder
+
+
+"""
+## Train the dual encoder model
+
+In this experiment, we freeze the base encoders for text and image, and make only
+the projection head trainable.
+"""
+
+projection_dims = 256
+num_projection_layers = 1
+dropout_rate = 0.1
+learning_rate = 0.003
+weight_decay = 0.0001
+num_epochs = 30
+batch_size = 256
+temperature = 0.07
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers, projection_dims, dropout_rate
+)
+vision_encoder.summary()
+
+text_encoder = create_text_encoder(num_projection_layers, projection_dims, dropout_rate)
+text_encoder.summary()
+
+optimizer = tfa.optimizers.AdamW(learning_rate=learning_rate, weight_decay=weight_decay)
+loss = ContrastiveSimilarityLoss(temperature)
+dual_encoder = create_dual_encoder(text_encoder, vision_encoder)
+dual_encoder.compile(optimizer=optimizer, loss=loss)
+dual_encoder.summary()
+
+"""
+Note that training the model with 60,000 image-caption pairs takes around 15 minutes
+per epoch using a P100 GPU accelerator"""
+
+print(f"Number of examples (caption-image pairs): {train_example_count}")
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {train_example_count // batch_size}")
+train_data = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+history = dual_encoder.fit(train_data, epochs=num_epochs)
+
+"""
+Plotting the training loss:
+"""
+
+plt.plot(history.history["loss"])
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.show()
+
+"""
+## Search for images using natural language
+
+The process of searching for images using using natural language consists of the
+following steps:
+1. Generating embeddings for the images by feeding them into the `vision_encoder`.
+2. Feed the natural language query to the `text_encoder` to generate a query embedding.
+3. Compute the similarity between the query embedding and the image embeddings
+in the index to retrieve the indices of the top matches.
+4. Lookup the paths of the top matching images to display them.
+
+Note that, after training the `dual encoder`, the only the fine-tuned `vision_encoder`
+and `text_encoder` models will be used, while the `dual_encoder` model will be discarded.
+"""
+
+"""
+### Generate embeddings for the images
+
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
+such as [Apache Spark](https://spark.apache.org) or [Apache Beam](https://beam.apache.org).
+Generating the image embeddings will take a few minutes.
+"""
+
+image_embeddings = []
+idx = 0
+batch_size = 512
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+while idx < len(image_paths):
```

Comment:
You can dramatically simplify this code by calling `vision_encoder.predict(dataset)` and putting your images in a dataset (could be done via `image_dataset_from_directory`)

You could also simply create a dataset from your image paths, then map `read_file`, `decode_jpeg`, `resize`, to the dataset, and batch it. Maybe than would be even simpler than `image_dataset_from_directory`.

Reply:
Done. Is what what you had in mind?